src/dodal/devices/areadetector/adaravis.py

```diff
@@ -1,99 +1,186 @@
+from ophyd_async.epics.areadetector.utils import (
+    ImageMode,
+    ad_r,
+    ad_rw,
+    stop_busy_record,
+)
+from ophyd_async.epics.areadetector.writers import HDFWriter, NDFileHDF, NDPluginStats
+
+
+class ADAravisTriggerMode(str, Enum):
+    """GigEVision GenICAM standard: on=externally triggered"""
+
+    on = "On"
+    off = "Off"
+
+
+class ADAravisTriggerSource(str, Enum):
+    # While not all enum elements may be physically supported by the hardware,
+    # DB record templates suggest they are valid options for underlying record
+    # cite: https://github.com/areaDetector/ADGenICam/tree/master/GenICamApp/Db
+    # (e.g. Mako G-125B TriggerSource has Line1-4)
+    freerun = "Freerun"
+    fixed_rate = "FixedRate"
+    software = "Software"
+    action_0 = "Action0"
+    action_1 = "Action1"
+    # Externally triggered on GPIO N
+    line_1 = "Line1"
+    line_2 = "Line2"
+    line_3 = "Line3"
+    line_4 = "Line4"
+
+
+_supported_external_triggers = (
+    DetectorTrigger.constant_gate,
+    DetectorTrigger.edge_trigger,
+)
+_csv_path = Path(__file__).parent / "adaravis_deadtime.csv"
+
+
+def _get_trigger_source(trigger: DetectorTrigger, gpio_number) -> ADAravisTriggerSource:
+    if trigger == DetectorTrigger.internal:
+        return ADAravisTriggerSource.fixed_rate
+    if trigger in _supported_external_triggers:
+        return ADAravisTriggerSource[f"line_{gpio_number}"]
+    raise ValueError("Unsupported trigger type: {trigger}")
+
+
+def _get_deadtime(model: str, pixel_format: Optional[str]) -> float:
+    # All version B, C so far have same deadtime as non-B/C
+    model = model.removesuffix("B").removesuffix("C")
+    with open(_csv_path) as csvfile:
+        reader = csv.DictReader(filter(lambda row: row[0] != "#", csvfile))
+        for row in reader:
+            if row["model"] == model and (
+                row.get(None) is None or pixel_format in row[None]
+            ):
+                return float(row["deadtime"])
+    return 0
+
+
+class AdAravisDriver(ADBase):
+    """Generic Driver supporting the Manta and Mako drivers.
+    Requires driver firmware up to date such that the Model_RBV is of the form "^(Mako|Manta) (model)$"
+    Attempts to fetch the deadtime from a lookup table in this directory on connect
+    """
+
+    def __init__(self, prefix: str, name: str = "") -> None:
+        self.trigger_mode = ad_rw(ADAravisTriggerMode, prefix + "TriggerMode")
+        self.trigger_source = ad_rw(ADAravisTriggerSource, prefix + "TriggerSource")
+        self.model = ad_r(str, prefix + "Model")
+        self.pixel_format = ad_r(str, prefix + "PixelFormat")
+        self.deadtime: Optional[float] = None
+        super().__init__(prefix, name=name)
+
+    async def connect(self, sim: bool = False, timeout: float = DEFAULT_TIMEOUT):
+        await super().connect(sim=sim, timeout=timeout)
+        model: str = await self.model.get_value()
+        pixel_format: str = await self.pixel_format.get_value()
```

Comment:
There's an implicit assumption here that the pixel format won't change at runtime (i.e. between `connect()` and the plan running). I guess that assumption _might_ be correct, but it would be worth making this more explicit. Or maybe recalculating the deadtime in `arm` or somewhere a bit closer to where it's actually used.

Reply:
Ideally it'd be in AdAravisDetector.prepare, as that's where it's used, arm is definitely closer. I'll see what I can do

Reply:
Moved to prepare